### PR TITLE
some SDC improvements

### PR DESCRIPTION
**Pull request by zingale**
## PR summary

* Implements reflecting BCs for PLM reconstruction with SDC (issue #652)

* Implements the density and temperature limits to burning for SDC (issue #598)

* Fixes the weighting of the flux registers for SDC (#658)

* Implements a well-balanced scheme for 2nd order PLM reconstruction (#294)

* Also adapts the bubble_convergence problem to work as a test for the well-balanced method and can now do second-order or fourth order initial model generation.

* Changes the meaning of the `plm_iorder` flag to now do 1 = piecewise constant, 2 = 2nd order MC limiter,  and >= 3 for 4th order MC limiter  

* Adds support for HSE boundary conditions in +X and +Y

## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite
- [ ] all functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
this also addresses #658

**Comment by zingale** [comment]
also addresses #294

**Comment by zingale** [comment]
All tests pass

File: Source/driver/_cpp_parameters [hunk 1]
```diff
@@ -104,7 +104,12 @@ ppm_predict_gammae           int           0                  y
 ppm_reference_eigenvectors   int           0                  y
 
 # for piecewise linear, reconstruction order to use
-plm_iorder                   int           2                  y
+# 1 = piecewise constant, 2 = MC limiter, >2 = 4th order MC limiter
+plm_iorder                   int           3                  y
```
Comment (maxpkatz) [inline Source/driver/_cpp_parameters]: I don't believe we should change the default PLM order, this is a breaking change without a good reason for it.

Comment (zingale) [reply Source/driver/_cpp_parameters]: I'm not changing it
I am just changing what the numbers mean.  So 3 is the new 2

Comment (zingale) [reply Source/driver/_cpp_parameters]: the idea is 1 is first order (piecewise constant), 2 is second order slopes, >=3 is the 4th order limited slopes

Comment (maxpkatz) [reply Source/driver/_cpp_parameters]: If someone has hardcoded plm_iorder in their inputs script, that inputs script will now do something different though, as a result of this change.

Comment (zingale) [reply Source/driver/_cpp_parameters]: yes, I know.  I am open to suggestions.  It just seemed odd to me to have 1 = first order, 2 = 4th order, 3 = 2nd order...

Comment (zingale) [reply Source/driver/_cpp_parameters]: okay, I've changed how this works.  `plm_iorder` remains the same as before now.  The choice of limiter when using linear slopes is now determined by `plm_limiter`.  I think this is actually a more natural way to do it, since both limiter choices are still linear slopes and therefore second-order.

File: Source/hydro/Castro_mol_hydro.cpp [hunk 2]
```diff
@@ -481,7 +486,14 @@ Castro::construct_mol_hydro_source(Real time, Real dt, MultiFab& A_update)
 	// Store the fluxes from this advance -- we weight them by the
 	// integrator weight for this stage
 
-        for (int idir = 0; idir < AMREX_SPACEDIM; ++idir) {
+        // For SDC, we store node 0 the only time we enter here (the
+        // first iteration) and we store the other nodes only on the
+        // last iteration.
+        if (time_integration_method == MethodOfLines ||
```
Comment (maxpkatz) [inline Source/hydro/Castro_mol_hydro.cpp]: This MethodOfLines integration method is now obsolete

File: Source/hydro/Castro_mol_nd.F90 [hunk 3]
```diff
@@ -59,66 +60,238 @@ subroutine ca_mol_plm_reconstruct(lo, hi, &
...
+
+       ! special care for reflecting BCs
+       if (idir == 1) then
+          if (lo(1) == domlo(1)) then
+
+             do k = lo(3)-dg(3), hi(3)+dg(3)
+                do j = lo(2)-dg(2), hi(2)+dg(2)
+
+                   ! reset the left state at domlo(1) if needed -- it is outside the domain
+                   if (physbc_lo(1) == Symmetry) then
```
Comment (maxpkatz) [inline Source/hydro/Castro_mol_nd.F90]: Perhaps we should do these if checks outside the loop?

Comment (zingale) [reply Source/hydro/Castro_mol_nd.F90]: I fixed this
